instructions/src/lib.rs

```diff
@@ -0,0 +1,162 @@
+pub enum MemSize {
+    Double,
+    Half
+}
+
+pub struct ProgramCodeBuilder {
+    instructions: Vec<u8>
+}
+
+impl <'p> ProgramCodeBuilder {
+
+    pub fn new() -> Self {
+        ProgramCodeBuilder { instructions: vec![] }
+    }
+
+    pub fn as_bytes(&self) -> &[u8] {
+        self.instructions.as_slice()
+    }
+
+    pub fn bind_value(&'p mut self) -> Operation<'p> {
+        Operation::new(self, 0xb7)
+    }
+
+    pub fn bind_register(&'p mut self) -> Operation<'p> {
+        Operation::new(self, 0xbf)
+    }
+
+    pub fn load_mem(&'p mut self, size: MemSize) -> Operation<'p> {
+        let opcode = match size {
+            MemSize::Double => 0x79,
+            MemSize::Half => 0x69
+        };
+        Operation::new(self, opcode)
+    }
+
+    pub fn exit(&'p mut self) -> Operation<'p> {
+        Operation::new(self, 0x95)
+    }
+}
+
+pub struct Operation<'p> {
+    program: &'p mut ProgramCodeBuilder,
+    op: u8,
+    dst: u8,
+    src: u8,
+    offset: u16,
+    immediate: u32
+}
+
+impl <'p> Operation<'p> {
+
+    fn new(program: &'p mut ProgramCodeBuilder, op: u8) -> Operation<'p> {
+        Operation {
+            program: program,
+            op: op,
+            dst: 0x00,
+            src: 0x00,
+            offset: 0x00_00,
+            immediate: 0x00_00_00_00
+        }
+    }
+
+    pub fn dst(&'p mut self, dst_register: u8) -> &'p mut Operation<'p> {
+        self.dst = dst_register;
+        self
+    }
+
+    pub fn src(&'p mut self, src_register: u8) -> &'p mut Operation<'p> {
+        self.src = src_register;
+        self
+    }
+
+    pub fn offset(&'p mut self, offset: u16) -> &'p mut Operation<'p> {
+        self.offset = offset;
+        self
+    }
+
+    pub fn value(&'p mut self, immediate: u32) -> &'p mut Operation<'p> {
+        self.immediate = immediate;
+        self
+    }
+
+    pub fn push(&'p mut self) -> &'p mut ProgramCodeBuilder {
+        {
+            let mut ins: &mut Vec<u8> = self.program.instructions.as_mut();
+            ins.push(self.op);
+            ins.push((self.src << 4) | self.dst);
+            ins.push((self.offset & 0x00_FF) as u8);
+            ins.push(((self.offset & 0xFF_00) >> 8) as u8);
+            ins.push((self.immediate & 0x00_00_00_FF) as u8);
+            ins.push(((self.immediate & 0x00_00_FF_00) >> 8) as u8);
+            ins.push(((self.immediate & 0x00_FF_00_00) >> 16) as u8);
+            ins.push(((self.immediate & 0xFF_00_00_00) >> 24) as u8);
+        }
+        self.program
+    }
+}
```

Comment:
I am not sure about that yet. I wanted to write a BPF program by calling chain of functions.
For instance:
```rust
program.bind_value().dst(0x00).value(0x00_00_11_22).push().bind_value().dst(0x01).value(0x11_22_00_00).push().add_registers().dst(0x00).src(0x01).push() ... /* and so forth */.exit()
```
Here I called function `push` referring _push on stack_, however I am not sure that it is a good name for it.

However, I have not considered to write it in this way:
```rust
program.add(Instruction::new(Class::Move).dst(0x00).src(0x01));
```